### PR TITLE
chore(MenuItem): Convert to RFC

### PR DESCRIPTION
**Pull request by assuncaocharles**
#### Pull request checklist

- [ ] Addresses an existing issue: Fixes #0000
- [ ] Include a change request file using `$ yarn change`

# BREAKING CHANGES

This PR converts `MenuItem` component to be functional. Restricting props set that will be passed to styles functions.

Related to #12237

## Prop sets

| `MenuItem`    |
| --------- |
| `primary` |
| `underlined` |
| `active` |
| `vertical` |
| `pointing` |
| `secondary` |
| `disabled` |
| `iconOnly` |
| `pills` |
| `icon` |
| `menu` |
| `inSubmenu` |
| `isFromKeyboard` |

#### Focus areas to test

(optional)

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/microsoft/fluentui/pull/13070)

### CODE REVIEW

File: packages/fluentui/react-northstar/src/themes/teams/components/Menu/menuItemStyles.ts [hunk 1]
```diff
@@ -436,7 +429,7 @@ const menuItemStyles: ComponentSlotStylesPrepared<MenuItemPropsAndState, MenuVar
       },
     },
 
-    ...(!!p.content && {
+    ...(!!p.hasContent && {
```
Comment (layershifter) [inline packages/fluentui/react-northstar/src/themes/teams/components/Menu/menuItemStyles.ts:432]:
```suggestion
    ...(p.hasContent && {
```

File: packages/fluentui/react-northstar/src/components/Menu/MenuItem.tsx [hunk 2]
```diff
@@ -146,238 +179,162 @@ export const menuItemSlotClassNames: MenuItemSlotClassNames = {
...
+      closeMenu: event => closeMenu(event),
+      closeMenuAndFocusTrigger: event => closeMenu(event, true),
+      doNotNavigateNextParentItem: event => {
+        event.stopPropagation();
+      },
+      closeAllMenus: event => closeAllMenus(event),
+    },
+    mapPropsToBehavior: () => ({
+      menuOpen,
+      menu: !!menu,
```
Comment (layershifter) [inline packages/fluentui/react-northstar/src/components/Menu/MenuItem.tsx:237]:
```suggestion
      hasMenu: !!menu,
```

File: packages/fluentui/react-northstar/src/components/Menu/MenuItem.tsx [hunk 3]
```diff
@@ -146,238 +179,162 @@ export const menuItemSlotClassNames: MenuItemSlotClassNames = {
...
+  const handleBlur = (e: React.SyntheticEvent) => {
+    setIsFromKeyboard(false);
 
-    _.invoke(this.props, 'onBlur', e, this.props);
+    _.invoke(props, 'onBlur', e, props);
   };
 
-  handleFocus = (e: React.SyntheticEvent) => {
-    this.setState({ isFromKeyboard: isFromKeyboard() });
+  const handleFocus = (e: React.SyntheticEvent) => {
```
Comment (layershifter) [inline packages/fluentui/react-northstar/src/components/Menu/MenuItem.tsx:320]:
```suggestion
  const handleFocus = (e: React.FocusEvent) => {
```

File: packages/fluentui/react-northstar/src/components/Menu/MenuItem.tsx [hunk 4]
```diff
@@ -146,238 +179,162 @@ export const menuItemSlotClassNames: MenuItemSlotClassNames = {
...
 
-    this.performClick(e);
-    _.invoke(this.props, 'onClick', e, this.props);
+    performClick(e);
+    _.invoke(props, 'onClick', e, props);
   };
 
-  handleBlur = (e: React.SyntheticEvent) => {
-    this.setState({ isFromKeyboard: false });
+  const handleBlur = (e: React.SyntheticEvent) => {
```
Comment (layershifter) [inline packages/fluentui/react-northstar/src/components/Menu/MenuItem.tsx:314]:
```suggestion
  const handleBlur = (e: React.FocusEvent) => {
```

File: packages/fluentui/react-northstar/src/components/Menu/MenuItem.tsx [hunk 5]
```diff
@@ -146,238 +179,162 @@ export const menuItemSlotClassNames: MenuItemSlotClassNames = {
...
+      styles,
+      variables,
+    }),
+    rtl: context.rtl,
+  });
+
+  const menuRef = React.createRef<HTMLElement>();
+  const itemRef = React.createRef<HTMLElement>();
+
+  const handleWrapperBlur = e => {
```
Comment (layershifter) [inline packages/fluentui/react-northstar/src/components/Menu/MenuItem.tsx:274]:
```suggestion
  const handleWrapperBlur = (e: React.FocusEvent) => {
```

File: packages/fluentui/react-northstar/src/components/Menu/MenuItem.tsx [hunk 6]
```diff
@@ -146,238 +179,162 @@ export const menuItemSlotClassNames: MenuItemSlotClassNames = {
...
-    closeMenuAndFocusTrigger: event => this.closeMenu(event, true),
-    doNotNavigateNextParentItem: event => {
-      event.stopPropagation();
-    },
-    closeAllMenus: event => this.closeAllMenus(event),
-  };
-
-  outsideClickHandler = e => {
-    if (!this.isSubmenuOpen()) return;
+  const outsideClickHandler = e => {
```
Comment (layershifter) [inline packages/fluentui/react-northstar/src/components/Menu/MenuItem.tsx:280]:
```suggestion
  const outsideClickHandler = (e: MouseEvent) => {
```

File: packages/fluentui/react-northstar/src/components/Menu/MenuItem.tsx [hunk 7]
```diff
@@ -146,238 +179,162 @@ export const menuItemSlotClassNames: MenuItemSlotClassNames = {
...
         e.stopPropagation();
         e.preventDefault();
       }
     }
   };
 
-  handleClick = (e: Event | React.SyntheticEvent) => {
-    const { disabled } = this.props;
-
+  const handleClick = (e: Event | React.SyntheticEvent) => {
```
Comment (layershifter) [inline packages/fluentui/react-northstar/src/components/Menu/MenuItem.tsx:304]:
```suggestion
  const handleClick = (e: Event | React.MouseEvent) => {
```

File: packages/fluentui/react-northstar/src/components/Menu/MenuItem.tsx [hunk 8]
```diff
@@ -146,238 +179,162 @@ export const menuItemSlotClassNames: MenuItemSlotClassNames = {
...
     ) {
-      this.trySetMenuOpen(false, e);
+      trySetMenuOpen(false, e);
     }
   };
 
-  performClick = e => {
-    const { active, menu } = this.props;
-
+  const performClick = e => {
```
Comment (layershifter) [inline packages/fluentui/react-northstar/src/components/Menu/MenuItem.tsx:290]:
```suggestion
  const performClick = (e: React.MouseEvent) => {
```

File: packages/fluentui/react-northstar/src/components/Menu/MenuItem.tsx [hunk 9]
```diff
@@ -146,238 +179,162 @@ export const menuItemSlotClassNames: MenuItemSlotClassNames = {
...
-    const { menu } = this.props;
-    const { menuOpen } = this.state;
-
+  const isSubmenuOpen = (): boolean => {
     return !!(menu && menuOpen);
   };
 
-  closeAllMenus = (e: Event) => {
-    if (!this.isSubmenuOpen()) {
+  const closeAllMenus = (e: React.SyntheticEvent) => {
```
Comment (layershifter) [inline packages/fluentui/react-northstar/src/components/Menu/MenuItem.tsx:330]:
```suggestion
  const closeAllMenus = (e: React.MouseEvent) => {
```

File: packages/fluentui/react-northstar/src/components/Menu/MenuItem.tsx [hunk 10]
```diff
@@ -387,16 +344,15 @@ class MenuItem extends AutoControlledComponent<WithAsProp<MenuItemProps>, MenuIt
     }
   };
 
-  closeMenu = (e: Event, forceTriggerFocus?: boolean) => {
-    if (!this.isSubmenuOpen()) {
+  const closeMenu = (e: React.SyntheticEvent, forceTriggerFocus?: boolean) => {
```
Comment (layershifter) [inline packages/fluentui/react-northstar/src/components/Menu/MenuItem.tsx:347]:
```suggestion
  const closeMenu = (e: React.MouseEvent, forceTriggerFocus?: boolean) => {
```

File: packages/fluentui/react-northstar/src/components/Menu/MenuItem.tsx [hunk 11]
```diff
@@ -405,33 +361,161 @@ class MenuItem extends AutoControlledComponent<WithAsProp<MenuItemProps>, MenuIt
     }
   };
 
-  openMenu = (e: Event) => {
-    const { menu } = this.props;
-    const { menuOpen } = this.state;
+  const openMenu = (e: React.SyntheticEvent) => {
```
Comment (layershifter) [inline packages/fluentui/react-northstar/src/components/Menu/MenuItem.tsx:364]:
```suggestion
  const openMenu = (e: React.MouseEvent) => {
```

File: packages/fluentui/react-northstar/src/components/Menu/MenuItem.tsx [hunk 12]
```diff
@@ -405,33 +361,161 @@ class MenuItem extends AutoControlledComponent<WithAsProp<MenuItemProps>, MenuIt
...
+      trySetMenuOpen(true, e);
+      _.invoke(props, 'onActiveChanged', e, { ...props, active: true });
       e.stopPropagation();
       e.preventDefault();
     }
   };
 
-  trySetMenuOpen(newValue: boolean, e: Event | React.SyntheticEvent, onStateChanged?: any) {
-    this.setState({ menuOpen: newValue });
+  const trySetMenuOpen = (newValue: boolean, e: Event | React.SyntheticEvent, onStateChanged?: any) => {
```
Comment (layershifter) [inline packages/fluentui/react-northstar/src/components/Menu/MenuItem.tsx:373]:
```suggestion
  const trySetMenuOpen = (newValue: boolean, e: MouseEvent | React.FocusEvent | React.Keyboard | React.MouseEvent, onStateChanged?: any) => {
```

**Comment by layershifter** [comment]
Let's avoid usages generic types like `Event` & `React.SyntheticEvent` and use exact types for handlers. It allows to better understand what happens in each handler and clarifies available properties.

**Comment by assuncaocharles** [comment]
> Let's avoid usages generic types like `Event` & `React.SyntheticEvent` and use exact types for handlers. It allows to better understand what happens in each handler and clarifies available properties.

Sure thing, thanks for point it out.

File: packages/fluentui/accessibility/src/behaviors/Menu/menuItemBehavior.ts [hunk 13]
```diff
@@ -73,7 +73,7 @@ export default menuItemBehavior;
 
 export type MenuItemBehaviorProps = {
   /** Indicated if menu item has submenu. */
-  menu?: boolean | object;
+  hasMenu?: boolean | object;
```
Comment (layershifter) [inline packages/fluentui/accessibility/src/behaviors/Menu/menuItemBehavior.ts:76]:
```suggestion
  hasMenu?: boolean;
```

File: packages/fluentui/react-northstar/src/components/Menu/MenuItem.tsx [hunk 14]
```diff
@@ -405,33 +361,165 @@ class MenuItem extends AutoControlledComponent<WithAsProp<MenuItemProps>, MenuIt
...
+  as: 'a',
+  accessibility: menuItemBehavior as Accessibility,
+  wrapper: { as: 'li' },
+  indicator: {},
+};
 
 MenuItem.create = createShorthandFactory({ Component: MenuItem, mappedProp: 'content' });
 
+// MenuItem.autoControlledProps = ['menuOpen'];
+
```
Comment (layershifter) [inline packages/fluentui/react-northstar/src/components/Menu/MenuItem.tsx:522]:
```suggestion
```

**Pull request merged**